virtual_rainforest/models/abiotic/radiation.py

```diff
@@ -0,0 +1,215 @@
+plant module. The implementation is based on David et al. (2017): Simple process-led
+algorithms for simulating habitats (SPLASH v.1.0): robust indices of radiation, evapo-
+transpiration and plant-available moisture, Geosci. Model Dev., 10, 689-708.
+
+Cloud cover and surface albedo also determine how much of the shortwave radiation that
+reaches the top of the canopy is reflected and how much remains to be absorbed via
+photosynthesis and re-emitted as longwave radiation by vegetation and forest floor.
+Scattering and re-absorption of longwave radiation are not considered at this stage.
+
+The radiation balance is calculated with the radiation_balance() function. At the
+moment, this happens at a daily timestep.
+
+# the following structural components are not implemented yet
+TODO include time dimension
+TODO logging, raise errors
+"""
+
+# from typing import Optional
+import numpy as np
+from numpy.typing import NDArray
+
+# from core.constants import CONSTANTS as C
+# this doesn't exist yet; optional scipy
+CLOUDY_TRANSMISSIVITY = 0.25
+"""cloudy transmittivity (Linacre, 1968)"""
+TRANSMISSIVITY_COEFFICIENT = 0.50
+"""angular coefficient of transmittivity (Linacre, 1968)"""
+FLUX_TO_ENERGY = 2.04
+"""from flux to energy conversion, umol/J (Meek et al., 1984)"""
+BOLZMAN_CONSTANT = 5.67e-8
+"""Stephan Bolzman constant W m-2 K-4"""
+SOIL_EMISSIVITY = 0.95
+"""Soil emissivity, default for tropical rainforest"""
+CANOPY_EMISSIVITY = 0.95
+"""canopy emissivity, default for tropical rainforest"""
+BEER_REGRESSION = 2.67e-5
+"""parameter in equation for atmospheric transmissivity based on regression of Beer’s
+radiation extinction function (Allen 1996)"""
+ALBEDO_VIS = np.array(0.03, dtype=np.float32)
+"""Albedo of visible light"""
+ALBEDO_SHORTWAVE = np.array(0.17, dtype=np.float32)
+"""Albedo of shortwave radiation"""
+CELSIUS_TO_KELVIN = 273.15
+"""factor to convert temperature in Celsius to absolute temperature in Kelvin"""
+SECOND_TO_DAY = 86400
+"""factor to convert between days and seconds."""
+
+
+class Radiation:
+    """Radiation balance."""
+
+    def __init__(self, elevation: NDArray[np.float32]) -> None:
+        """Initializes point-based radiation method."""
+
+        self.elevation = elevation
+        """Elevation above sea level, [m]"""
+
+        # populated later
+        self.ppfd = NDArray[np.float32]
+        """Daily top of canopy photosynthetic photon flux density, [mol m-2]"""
+        self.topofcanopy_radiation = NDArray[np.float32]
+        """Daily top of canopy downward shortwave radiation, [J m-2]"""
+        self.longwave_canopy = NDArray[np.float32]
+        """Daily longwave radiation from n individual canopy layers, [J m-2]"""
+        self.longwave_soil = NDArray[np.float32]
+        """Daily longwave radiation from soil, [J m-2]"""
+        self.netradiation_surface = NDArray[np.float32]
+        """Daily net shortwave radiation at the surface (= forest floor), [J m-2]"""
+
+    def calc_ppfd(
+        self,
```

Comment:
Actually, I think this is important. `Radiation` as a class is currently only intended ever to be used as:

```
r = Radiation(elev)
r.radiation_balance(lots_of_args)
```

If you try and use the other methods, it all works _provided_ you run them in the right order and pass `tau` correctly. If `Radiation` is basically a recipe for applying a set of radiation calculations in the right order, then making those `calc_` functions external and only providing `Radiation.radiation_balance` ensures that the class can't be used incorrectly.

I don't know if there is stuff to be added to `Radiation` but it could be that all of the `radiation_balance` code can be moved into `__init__` and then you can just use:

```
r = Radiation(elev, lots_of_args)
```

That doesn't make sense if there are other things to add in though.

Reply:
Thanks @davidorme for this suggestion. I agree that this is an important point, something I was trying to get to some time ago when we had the meetings on overall model structure, but I couldn't articulate it very well without the examples to show. If the concept of a class is not appropriate for the microclimate (this might not only affect radiation but other components, too), I'm happy to change it. What do you think @alexdewar  and @dalonsoa ?

Reply:
Just to add that I think a class is right here. It acts as a convenient container for multiple variables calculated from the radiation inputs and you want to keep those packaged together.

It could be that `Radiation` makes most sense as a [`dataclass`](https://docs.python.org/3/library/dataclasses.html). This is a kind of stripped down syntax to create classes that at their simplest just hold data. That would package up all of the `lots_of_args` into one block and then the `__post_init__` method can be used to calculate the rest. That really would be very little different from what you have now, but it is just a cleaner syntax.

Reply:
This is a mangled example because I'm not using the right inputs etc... but:

```python
@dataclass
class Radiation():
    elev: NDArray
    other_init_args: NDArray
    ppfd: NDArray = field(init=False)

    def __post_init__(self):
        self.ppfd = calc_ppfd(self.other_init_args)
```

And then you get `Radiation(elev, other_init_args)` as the class signature and `__post_init__` is basically your `radiation_balance` method.

Reply:
I agree with @alexdewar and @davidorme comments. Making things more functional by extracting some of these methods as functions will simplify the testing and maintenance of the code, as well as the chances of using it wrong. I will add using the full `calculate_ppfd`. You're not saving much space and it makes the name of the function readable.

Reply:
Thank you both, I'll change the structure to have more pure functions.
About the values to be stored, most of the attributes are inputs to other modules (for example PPFD is a key input for the plants, net_radiation is the main input to the energy balance) and I would ideally like to store them in our central data object. This is where the top down and bottom up clashes a little bit (in my brain at least), I didn't fully understand the discussion we had around the Modelclass/loading the data object/inheritence/overwriting/timing...When I first drafted this I only thought about the functions without the VR context, now I will need to fit this in the overall structure.